src/Features/Core/Portable/AddDebuggerDisplay/AbstractAddDebuggerDisplayCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.AddDebuggerDisplay
+{
+    internal abstract class AbstractAddDebuggerDisplayCodeRefactoringProvider<TTypeDeclarationSyntax, TMethodDeclarationSyntax> : CodeRefactoringProvider
+        where TTypeDeclarationSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+    {
+        private static readonly SyntaxAnnotation s_trackingAnnotation = new SyntaxAnnotation();
+
+        private const string DebuggerDisplayMethodName = "GetDebuggerDisplay";
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var type =
+                await context.TryGetRelevantNodeAsync<TTypeDeclarationSyntax>().ConfigureAwait(false)
+                ?? await GetRelevantTypeFromMethodAsync(context).ConfigureAwait(false);
+
+            if (type is null) return;
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var typeSymbol = (ITypeSymbol)semanticModel.GetDeclaredSymbol(type, context.CancellationToken);
+
+            if (IsClassOrStruct(typeSymbol) && !HasDebuggerDisplayAttribute(typeSymbol, semanticModel.Compilation))
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Add_DebuggerDisplay,
+                    cancellationToken => ApplyAsync(context.Document, type, cancellationToken)));
+            }
+        }
+
+        private static async Task<TTypeDeclarationSyntax> GetRelevantTypeFromMethodAsync(CodeRefactoringContext context)
+        {
+            var method = await context.TryGetRelevantNodeAsync<TMethodDeclarationSyntax>().ConfigureAwait(false);
+            if (method != null)
+            {
+                var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+                var methodSymbol = (IMethodSymbol)semanticModel.GetDeclaredSymbol(method);
+
+                if (IsToStringOverride(methodSymbol) || IsDebuggerDisplayMethod(methodSymbol))
+                {
+                    return method.FirstAncestorOrSelf<TTypeDeclarationSyntax>();
+                }
+            }
+
+            return null;
+        }
+
+        private static bool IsToStringOverride(IMethodSymbol methodSymbol)
```

Comment:
while your check is fine.  i would have actually preferred a check that said that we walked up (through .OverriddenMethod) up to `object.ToString`.

Reply:
I think I'm going the opposite direction and will drop the override check since that would be more consistent with the behavior of `GetDebuggerDisplay() => ToString();`.

We could gen `GetDebuggerDisplay() => ((object)this).ToString();` but at that point it's probably not what the dev wants anyway.

Reply:
I think I'm going the opposite direction and will drop the override check since that would be more consistent with the behavior of `GetDebuggerDisplay() => ToString();`.

We could gen `GetDebuggerDisplay() => ((object)this).ToString();` but at that point it's probably not what the dev wants anyway.

Reply:
I think I'm going the opposite direction and will drop the override check since that would be more consistent with the behavior of `GetDebuggerDisplay() => ToString();`.

We could gen `GetDebuggerDisplay() => ((object)this).ToString();` but at that point it's probably not what the dev wants anyway.